[1.14 regression] truthy check for enums

**Bug Report**

It appears the truthy check for enums is incorrect in 1.14 when `warn_unreachable` is set to True.

For example

```python
import enum


class MyEnum(enum.StrEnum):
    EMPTY = ""
    NOT_EMPTY = "asdf"


def is_truthy(e: MyEnum) -> None:
    if e:
        print(f"{e} was truthy: {len(e)}")
    else:
        print(f"{e} was falsey: {len(e)}")


is_truthy(MyEnum.EMPTY)
is_truthy(MyEnum.NOT_EMPTY)
```

With a `mypy.ini`

```ini
[mypy]
warn_unreachable = True
```

If we run the script we get:

```
 was falsey: 0
asdf was truthy: 4
```

Which shows the behaviour of these enums is such that they can be used in an if condition

but running `mypy a.py` results in:

```
a.py:13: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.14
- Python version used: tested with and saw same problem on python3.12 and python3.13

--- a/mypy/typeops.py
+++ b/mypy/typeops.py
@@ -658,9 +658,6 @@ def _get_type_method_ret_type(t: Type, *, name: str) -> Type | None:
 
     if isinstance(t, Instance):
         sym = t.type.get(name)
-        # Fallback to the metaclass for the lookup when necessary
-        if not sym and (m := t.type.metaclass_type):
-            sym = m.type.get(name)
         if sym:
             sym_type = get_proper_type(sym.type)
             if isinstance(sym_type, CallableType):
@@ -733,7 +730,10 @@ def false_only(t: Type) -> ProperType:
         if ret_type:
             if not ret_type.can_be_false:
                 return UninhabitedType(line=t.line)
-        elif isinstance(t, Instance) and t.type.is_final:
+        elif isinstance(t, Instance):
+            if t.type.is_final or t.type.is_enum:
+                return UninhabitedType(line=t.line)
+        elif isinstance(t, LiteralType) and t.is_enum_literal():
             return UninhabitedType(line=t.line)
 
         new_t = copy_type(t)

--- a/mypy/typeops.py
+++ b/mypy/typeops.py
@@ -648,9 +648,7 @@ def _remove_redundant_union_items(items: list[Type], keep_erased: bool) -> list[
     return items
 
 
-def _get_type_method_ret_type(t: Type, *, name: str) -> Type | None:
-    t = get_proper_type(t)
-
+def _get_type_method_ret_type(t: ProperType, *, name: str) -> Type | None:
     # For Enum literals the ret_type can change based on the Enum
     # we need to check the type of the enum rather than the literal
     if isinstance(t, LiteralType) and t.is_enum_literal():

--- a/test-data/unit/check-enum.test
+++ b/test-data/unit/check-enum.test
@@ -181,27 +181,100 @@ def infer_truth(truth: Truth) -> None:
 [case testEnumTruthyness]
 # mypy: warn-unreachable
 import enum
+from typing_extensions import Literal
+
 class E(enum.Enum):
-    x = 0
-if not E.x:
-    "noop"
+    zero = 0
+    one = 1
+
+def print(s: str) -> None: ...
+
+if E.zero:
+    print("zero is true")
+if not E.zero:
+    print("zero is false")  # E: Statement is unreachable
+
+if E.one:
+    print("one is true")
+if not E.one:
+    print("one is false")  # E: Statement is unreachable
+
+def main(zero: Literal[E.zero], one: Literal[E.one]) -> None:
+    if zero:
+        print("zero is true")
+    if not zero:
+        print("zero is false")  # E: Statement is unreachable
+    if one:
+        print("one is true")
+    if not one:
+        print("one is false")  # E: Statement is unreachable
 [builtins fixtures/tuple.pyi]
-[out]
-main:6: error: Statement is unreachable
 
 [case testEnumTruthynessCustomDunderBool]
 # mypy: warn-unreachable
 import enum
 from typing_extensions import Literal
+
 class E(enum.Enum):
-    x = 0
+    zero = 0
+    one = 1
     def __bool__(self) -> Literal[False]:
         return False
-if E.x:
-    "noop"
+
+def print(s: str) -> None: ...
+
+if E.zero:
+    print("zero is true")  # E: Statement is unreachable
+if not E.zero:
+    print("zero is false")
+
+if E.one:
+    print("one is true")  # E: Statement is unreachable
+if not E.one:
+    print("one is false")
+
+def main(zero: Literal[E.zero], one: Literal[E.one]) -> None:
+    if zero:
+        print("zero is true")  # E: Statement is unreachable
+    if not zero:
+        print("zero is false")
+    if one:
+        print("one is true")  # E: Statement is unreachable
+    if not one:
+        print("one is false")
 [builtins fixtures/enum.pyi]
-[out]
-main:9: error: Statement is unreachable
+
+[case testEnumTruthynessStrEnum]
+# mypy: warn-unreachable
+import enum
+from typing_extensions import Literal
+
+class E(enum.StrEnum):
+    empty = ""
+    not_empty = "asdf"
+
+def print(s: str) -> None: ...
+
+if E.empty:
+    print("empty is true")
+if not E.empty:
+    print("empty is false")
+
+if E.not_empty:
+    print("not_empty is true")
+if not E.not_empty:
+    print("not_empty is false")
+
+def main(empty: Literal[E.empty], not_empty: Literal[E.not_empty]) -> None:
+    if empty:
+        print("empty is true")
+    if not empty:
+        print("empty is false")
+    if not_empty:
+        print("not_empty is true")
+    if not not_empty:
+        print("not_empty is false")
+[builtins fixtures/enum.pyi]
 
 [case testEnumUnique]
 import enum

--- a/test-data/unit/fixtures/enum.pyi
+++ b/test-data/unit/fixtures/enum.pyi
@@ -11,6 +11,8 @@ class tuple(Generic[T]):
     def __getitem__(self, x: int) -> T: pass
 
 class int: pass
-class str: pass
+class str:
+    def __len__(self) -> int: pass
+
 class dict: pass
 class ellipsis: pass

